### PR TITLE
Ayjohnso/create information barriers from aus and sgs non sds

### PR DESCRIPTION
**Pull request by AyronJohnson**
Script to create IB's from administrative units and security groups for customers not using SDS.

### CODE REVIEW

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 1]
```diff
@@ -0,0 +1,316 @@
+<#
+.SYNOPSIS
+This script is designed to create Information Barrier Policies for each Administrative Units and Security Group not created by SDS from an O365 tenant.
+
+.DESCRIPTION
+This script will read from Azure, and output the administrative units and security groups to CSVs.  Afterwards, you are prompted to confirm that you want to create the organization segments needed, then create and apply the information barrier policies.  A folder will be created in the same directory as the script itself and contains a log file which details the organization segments and information barrier policies created.  The rows of the csv files can be reduced to only target specific administrative units and security groups.  Nextlink in the log can be used for the skipToken script parameter to continue where the script left off in case it does not finish. 
+
+.EXAMPLE
+PS> .\Create-IBs_from_AUs_and_SGs_non_SDS.ps1
```
Comment (acherrymsft) [inline CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:9]: Nit: Name is not very user friendly.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+This script is designed to create Information Barrier Policies for each Administrative Units and Security Group not created by SDS from an O365 tenant.
+
+.DESCRIPTION
+This script will read from Azure, and output the administrative units and security groups to CSVs.  Afterwards, you are prompted to confirm that you want to create the organization segments needed, then create and apply the information barrier policies.  A folder will be created in the same directory as the script itself and contains a log file which details the organization segments and information barrier policies created.  The rows of the csv files can be reduced to only target specific administrative units and security groups.  Nextlink in the log can be used for the skipToken script parameter to continue where the script left off in case it does not finish. 
+
+.EXAMPLE
+PS> .\Create-IBs_from_AUs_and_SGs_non_SDS.ps1
+
+.NOTES
+This script uses features required by Information Barriers version 3 or above enabled in your tenant.  Existing Organization Segments and Information Barriers created by a legacy version should be removed prior to upgrading.
```
Comment (acherrymsft) [inline CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:12]: First sentence doesn't seem quite right. Maybe:
This script uses features that require Information Barriers version 3 or above to be enabled in your tenant.

There are other options.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+#>
+
+Param (
+    [Parameter(Mandatory=$false)]
+    [string] $skipToken= ".",
```
Comment (acherrymsft) [inline CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:17]: Default to ".", not blank?

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    [string] $outFolder = ".\non_SDS_InformationBarriers",
+    [Parameter(Mandatory=$false)]
+    [string] $graphVersion = "beta",
+    [switch] $PPE = $false
+)
+
+$GraphEndpointProd = "https://graph.microsoft.com"
+$GraphEndpointPPE = "https://graph.microsoft-ppe.com"
+
+function Get-PrerequisiteHelp
```
Comment (acherrymsft) [inline CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:28]: Why not just have this as standard doc comments?

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+{
+    Write-Output @"
+========================
+ Required Prerequisites
+========================
+
+1. This script uses features required by Information Barriers version 3 or above enabled in your tenant.
```
Comment (acherrymsft) [inline CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:35]: Same sentence as earlier.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+$logFilePath = "$outFolder\create_non_SDS_InformationBarriers.log"
+$csvFilePathAU = "$outFolder\AdministrativeUnits.csv"
+$csvFilePathSG = "$outFolder\SecurityGroups.csv"
+
+# List used to request access to data
+$graphScopes = "AdministrativeUnit.ReadWrite.All, Group.ReadWrite.All, Directory.ReadWrite.All"
+
+try 
+{
+    Import-Module Microsoft.Graph.Authentication -MinimumVersion 0.9.1 | Out-Null
```
Comment (acherrymsft) [inline CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:258]: Nit: Out-Error?

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+ }
+
+Write-Host "`nActivity logged to file $logFilePath `n" -ForegroundColor Green
+
+Connect-Graph -scopes $graphScopes | Out-Null
+Connect-IPPSSession | Out-Null
+
+Get-AdministrativeUnits
+
+Write-Host "`nYou are about to create organization segments and information barrier policies from administrative units. `nIf you want to skip any administrative units, edit the file now and remove the corresponding lines before proceeding. `n" -ForegroundColor Yellow
```
Comment (acherrymsft) [inline CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:291]: It seems like being able to exit and edit the file would be more elegant than "do this in the background, then hit to continue". If this meets the business needs, I won't block.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+        $grpCtr = 0 # Counter for security groups retrieved
+        
+        foreach ($grp in $grps){
+            if ( !($grp.extension_fe2174665583431c953114ff7268b7b3_Education_ObjectType) ) { # Filtering out security groups already with SDS attribute *note: SG's don't have sync source attribute
+                $grpList += [pscustomobject]@{"GroupObjectId"=$grp.Id;"GroupDisplayName"=$grp.DisplayName}
+                $grpCtr++
+            }
+        }
+
+        $grpList | Export-Csv $csvFilePathSG -Append -NotypeInformation
```
Comment (acherrymsft) [inline CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:191]: Nit: Inconsistent capitalization on NoTypeInformation

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+4.  Ensure that All Teachers security group is enabled in SDS and exists in Azure Active Directory.
+
+5.  Retry this script.  If you still get an error about failing to load the Microsoft Graph module, troubleshoot why "Import-Module Microsoft.Graph.Authentication -MinimumVersion 0.9.1" isn't working and do the same for the Exchange Online Management Module.
+
+(END)
+========================
+"@
+}
+
+function Get-AdministrativeUnits {
```
Comment (acherrymsft) [inline CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:60]: The large degree of similarity between the AU and SG functionality here really seems like there should be more code shared.

**Review by acherrymsft — COMMENTED** [review]
A few neatness things, but it really seems like the AU and SG code are so similar that they shouldn't be separate functions.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+            $lastRefreshed = Get-Date
+        }
+
+        $auResponse = Invoke-GraphRequest -Uri $auUri -Method GET
+        $aus = $auResponse.value
+        
+        $auCtr = 0 # Counter for AUs retrieved
+        
+        foreach ($au in $aus){
+            if ( $au.extension_fe2174665583431c953114ff7268b7b3_Education_SyncSource -eq $null ) { # Filtering out AU already with SDS attribute
```
[the remark this answers is not in the document]
Comment (AyronJohnson) [reply CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:95]: @acherrymsft Yes, are you suggesting to use a specific one?  There can be different types that is why I chose sync source.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 (hunk 6 above)
Comment (AyronJohnson) [reply CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:258]: I have write-error on the catch.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 (hunk 4 above)
Comment (AyronJohnson) [reply CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:28]: Moving to .Notes

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 (hunk 1 above)
Comment (AyronJohnson) [reply CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:9]: @acherrymsft I was thinking of expanding the acronyms, but that leads to a very long file name.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 (hunk 7 above)
Comment (AyronJohnson) [reply CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1]: Added logic to go straight to the IB creation if user exists script prior.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 (hunk 9 above)
Comment (AyronJohnson) [reply CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1]: Combined the AU and SG functions.

**Comment by AyronJohnson** [comment]
@acherrymsft see my responses and changes based on your comments.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 (hunk 1 above)
Comment (acherrymsft) [reply CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1]: create-nonSDS-IBs ?

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+    d. If you are returned to the PowerShell session without error, you are correctly set up.
+
+4.  Ensure that All Teachers security group is enabled in SDS and exists in Azure Active Directory.
+
+5.  Retry this script.  If you still get an error about failing to load the Microsoft Graph module, troubleshoot why "Import-Module Microsoft.Graph.Authentication -MinimumVersion 0.9.1" isn't working and do the same for the Exchange Online Management Module.
+#>
+
+Param (
+    [Parameter(Mandatory=$false)]
+    [string] $skipToken= "",
```
Comment (acherrymsft) [inline CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:59]: Q: Can you just use $null? I hadn't thought of it previously -- and am okay with the current implementation.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 (hunk 10 above)
[the remark this answers is not in the document]
Comment (acherrymsft) [reply CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1]: My goal was to enable merging the two code paths, which you already did.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 (hunk 11 above)
Comment (AyronJohnson) [reply CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1:59]: Tried that before but causes failure.

File: CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1 (hunk 1 above)
Comment (AyronJohnson) [reply CustomSolutions/Create-IBs_from_AUs_and_SGs_non_SDS.ps1]: Sure, I guess I was trying to be specific as to from what the IB's are created.

**Pull request merged**